homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,175 @@
+import logging
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME, CONF_ENTITY_ID, ATTR_UNIT_OF_MEASUREMENT)
+from homeassistant.core import callback
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.filter import (
+    Filter, FILTER_LOWPASS, FILTER_OUTLIER)
+from homeassistant.helpers.event import async_track_state_change
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_FILTER = 'filter'
+ATTR_PRE_FILTER = 'pre_filter_state'
+
+CONF_FILTER_OPTIONS = 'options'
+CONF_FILTER_NAME = 'filter'
+
+TYPE_LOWPASS = 'lowpass'
+TYPE_OUTLIER = 'outlier'
+
+FILTER_MAP = {
+             TYPE_LOWPASS: FILTER_LOWPASS,
+             TYPE_OUTLIER: FILTER_OUTLIER,
+             }
+
+DEFAULT_NAME_TEMPLATE = "{} filter {}"
+ICON = 'mdi: chart-line-variant'
+
+# ALL filter arguments must be OPTIONAL
+FILTER_SCHEMA = vol.Schema({
+    vol.Optional('time_constant'): vol.Coerce(int),
+    vol.Optional('constant'): vol.Coerce(int)
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ENTITY_ID): cv.entity_id,
+    vol.Required(CONF_FILTER_NAME):
+        vol.Any(TYPE_LOWPASS, TYPE_OUTLIER),
+    vol.Optional(CONF_FILTER_OPTIONS): FILTER_SCHEMA,
+    vol.Optional(CONF_NAME, default=None): cv.string,
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Filter sensor."""
+    entity_id = config.get(CONF_ENTITY_ID)
+    filter_name = config.get(CONF_FILTER_NAME)
+    name = config.get(CONF_NAME)
+    if name is None:
+        name = DEFAULT_NAME_TEMPLATE.format(entity_id, filter_name)
+
+    async_add_devices([
+            FilterSensor(hass, name, entity_id, filter_name,
+                         config.get(CONF_FILTER_OPTIONS, dict()))
+        ], True)
+    return True
+
+
+class FilterSensor(Entity):
+    """Representation of a Filter sensor."""
+
+    def __init__(self, hass, name, entity_id, filter_name, filter_args):
+        """Initialize the Filter sensor."""
+        self._name = name
+        self._filter_name = filter_name
+        self._unit_of_measurement = None
+        self._pre_filter_state = self._state = None
+
+        self._filterdata = self.filterdata_factory(FILTER_MAP[filter_name],
+                                                   **filter_args)()
+
+        @callback
+        # pylint: disable=invalid-name
+        def async_stats_sensor_state_listener(entity, old_state, new_state):
+            """Handle the sensor state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+
+            try:
+                self._pre_filter_state = new_state.state
+                self._filterdata.update(self._pre_filter_state)
+            except ValueError:
+                _LOGGER.error("This component can only filter numbers")
```

Comment:
I know this is a RFC and the following comment is quite specific 😀: I have sensors that periodically spit out `STATE_UNKNOWN` states, as I believe quite a few platforms do (MQTT, WUnderground, ). If I'm correct this would result in a complete spam of the logger due to this `_LOGGER.error`.

I understand it would be good to tell users that only numbers can be filtered, but 1. that could spam the logs and 2. I think users are smart enough to know that only numbers can be filtered (maybe I'm too naïve about this though...)

Also: Your code does this correctly, but writing this just in case: we should make absolutely sure that filters only spit out values when their "parent" also returns a value - otherwise this could result in the recorder+history_graph becoming unusable (from experience 😄)

Reply:
Well, after the stuff I've seen in a Portuguese Home Assistant support group, I can assure you that people are that naive... :(

But the point you made about STATE_UNKNOW / None is quite write (updating the code)

Reply:
Perhaps it would be better to display in the attributes, some useful statistics about how well the filter is performing. E.g. fraction of outliers = 10%.  

Typically, when your component creates errors, people start threads to complain about it so better to avoid :-)